subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+
+	zms_ate_crc8_update(&gc_done_ate);
+
+	return zms_flash_ate_wrt(fs, &gc_done_ate);
+}
+
+static int zms_add_empty_ate(struct zms_fs *fs, uint64_t addr)
+{
+	struct zms_ate empty_ate;
+	size_t ate_size;
+	uint8_t cycle_cnt;
+	int rc = 0;
+
+	addr &= ADDR_SECT_MASK;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	LOG_DBG("Adding empty ate at %llx", (uint64_t)(addr + fs->sector_size - 2 * ate_size));
+	empty_ate.id = ZMS_HEAD_ID;
+	empty_ate.len = 0xffff;
+	empty_ate.offset = 0U;
+	empty_ate.metadata = ZMS_DEFAULT_VERSION;
+
+	rc = zms_get_sector_cycle(fs, addr, &cycle_cnt);
+	if (rc == -ENOENT) {
+		/* sector never used */
+		cycle_cnt = 0;
+	} else if (rc) {
+		/* bad flash read */
+		return rc;
+	}
+
+	/* increase cycle counter */
+	empty_ate.cycle_cnt = (cycle_cnt + 1) % BIT(8);
+	zms_ate_crc8_update(&empty_ate);
+	fs->ate_wra = addr + fs->sector_size - (2 * ate_size);
+
+	return zms_flash_ate_wrt(fs, &empty_ate);
+}
+
+static int zms_get_sector_cycle(struct zms_fs *fs, uint64_t addr, uint8_t *cycle_cnt)
+{
+	int rc;
+	size_t ate_size;
+	struct zms_ate empty_ate;
+	uint64_t empty_addr;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	empty_addr = addr & ADDR_SECT_MASK;
+	empty_addr += fs->sector_size - 2 * ate_size;
+
+	/* read the cycle counter of the current sector */
+	rc = zms_flash_ate_rd(fs, empty_addr, &empty_ate);
+	if (rc < 0) {
+		/* flash error */
+		return rc;
+	}
+
+	if (zms_empty_ate_valid(fs, &empty_ate)) {
+		*cycle_cnt = empty_ate.cycle_cnt;
+	} else {
+		/* there is no empty ATE in this sector */
+		return -ENOENT;
+	}
+
+	return 0;
+}
+
+static int zms_get_sector_header(struct zms_fs *fs, uint64_t addr, struct zms_ate *empty_ate,
+				 struct zms_ate *close_ate)
+{
+	int rc;
+	size_t ate_size;
+	uint64_t close_addr;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	close_addr = addr & ADDR_SECT_MASK;
+	close_addr += fs->sector_size - ate_size;
+
+	/* read the first ate in the sector to get the close ATE */
+	rc = zms_flash_ate_rd(fs, close_addr, close_ate);
+	if (rc) {
+		return rc;
+	}
+
+	/* read the second ate in the sector to get the empty ATE */
+	rc = zms_flash_ate_rd(fs, close_addr - ate_size, empty_ate);
+	if (rc) {
+		return rc;
+	}
+
+	return 0;
+}
+
+/* garbage collection: the address ate_wra has been updated to the new sector
+ * that has just been started. The data to gc is in the sector after this new
+ * sector.
+ */
+static int zms_gc(struct zms_fs *fs)
+{
+	int rc, previous_sector_num = -1;
+	struct zms_ate close_ate, gc_ate, wlk_ate, empty_ate;
+	uint64_t sec_addr, gc_addr, gc_prev_addr, wlk_addr, wlk_prev_addr, data_addr, stop_addr,
+		previous_ate_wra;
+	uint8_t current_cycle, previous_cycle = 0;
+	size_t ate_size;
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	rc = zms_get_sector_cycle(fs, fs->ate_wra, &fs->sector_cycle);
+	if (rc == -ENOENT) {
```

Comment:
if the check would be replaced with
```
if (rc !=0 && != -ENOENT) {
   return rc;
```
you can get the code more streamlined as the entire block for `rc == -ENOENT` gets moved left.

Reply:
I do not agree here.
The else for your statement is : (rc == 0) or (rc == -ENOENT) and both cases are different so we won't move any code to the left